addons/mass_mailing/controllers/main.py

```diff
@@ -21,7 +21,14 @@ def mailing(self, mailing_id, email=None, res_id=None, **post):
                 res_ids = contacts.ids
             else:
                 res_ids = [res_id]
-            mailing.update_opt_out(mailing_id, email, res_ids, True)
+
+            # Do not save unsubscription if there is any failure or flaw
+            with request.env.cr.savepoint():
+                mailing.update_opt_out(mailing_id, email, res_ids, True)
+                right_token = mailing._unsubscribe_token(res_id, email)
+                if token != right_token:
+                    raise exceptions.AccessDenied()
```

Comment:
This whole this with a savepoint seems useless. If you raise an exception the transaction will be rolled back, so the unsubscription won't be saved. Savepoints should only be used (exceptionally) for sub-transactions.
Also, shouldn't you be testing the token before calling `update_opt_out()`, so there is actually nothing to rollback?